test/watcher.test.js

```diff
@@ -0,0 +1,140 @@
+const path = require('path');
+const ncc = global.coverage ? require("../src/index") : require("../");
+const fs = require('fs');
+
+// Based on the NodeWatchFileSystem class at:
+// - https://github.com/webpack/webpack/blob/master/lib/node/NodeWatchFileSystem.js
+// which in turn exposes:
+// - https://www.npmjs.com/package/watchpack
+
+class CustomWatchFileSystem {
+  constructor(watchStart, watchEnd) {
+    this.closed = false;
+    // paused allows the watchers to stay open for the next build
+    this.paused = false;
+    this.changeCallback = undefined;
+    this.watchStart = watchStart;
+    this.watchEnd = watchEnd;
+
+    // Webpack requires us to track this stuff
+    this.files = undefined;
+    this.dirs = undefined;
+    this.missing = undefined;
+    this.timestamps = new Map();
+
+    // this will be populated for us by ncc
+    this.inputFileSystem = undefined;
+  }
+  
+  triggerChanges (changed, removed) {
+    if (!this.paused) {
+      const newTime = +Date.now();
+      for (const file of changed)
+        this.timestamps.set(file, {
+          safeTime: newTime + 10,
+          accuracy: 10,
+          timestamp: newTime
+        });
+      for (const file of removed)
+        this.timestamps.set(file, null);
+
+      for (const file of changed)
+        this.inputFileSystem.purge(file);
+      for (const file of removed)
+        this.inputFileSystem.purge(file);
+
+      this.changeCallback(
+        null,
+        this.timestamps,
+        this.timestamps,
+        removed
+      );
+    }
+  }
+
+  // This is called on every rebuild
+  watch (files, dirs, missing, startTime, options, changeCallback) {
+    this.files = new Set(files);
+    this.dirs = new Set(dirs);
+    this.missing = new Set(missing);
+
+    // empty object indicates "unknown" timestamp
+    // (that is, not cached)
+    for (const item of files)
+      this.timestamps.set(item, {});
+    for (const item of dirs)
+      this.timestamps.set(item, {});
+    // null represents "no file"
+    for (const item of missing)
+      this.timestamps.set(item, null);
+
+    this.paused = false;
+    this.changeCallback = changeCallback;
+
+    // ...Start watching files, dirs, mising
+    setImmediate(() => {
+      this.watchStart(files, dirs, missing);
+    });
+		
+    return {
+      close: () => {
+        this.watchEnd();
+      },
+      pause: () => {
+        this.paused = true;
+      },
+      getFileTimestamps: () => {
+        return this.timestamps;
+      },
+      getContextTimestamps: () => {
+        return this.timestamps;
+      }
+    };
+  }
+}
+
+jest.setTimeout(20000);
+
+it('Should support custom watch API', async () => {
+  let buildCnt = 0;
+  const buildFile = path.resolve('./test/integration/twilio.js');
+  await new Promise((resolve, reject) => {
+    const watcher = new CustomWatchFileSystem(function watchStart (files, dirs, missing) {
```

Comment:
This is a webpack WatchFileSystem class API, which we can pass to the ncc `watch` option.